flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/HsFileDataManagerTest.java

```diff
@@ -340,6 +341,57 @@ void testRegisterSubpartitionReaderAfterReleased() {
                 .hasMessageContaining("HsFileDataManager is already released.");
     }
 
+    /**
+     * When the result partition fails, the view lock may be obtained when the FileDataManager lock
+     * is held. In the same time, the downstream thread will acquire the lock of the FileDataManager
+     * when acquiring the view lock. To avoid this deadlock, the logical of subpartition view
+     * release subpartition reader and subpartition reader fail should not be inside lock.
+     */
+    @Test
+    void testDeadLock() throws Exception {
+        TestingHsSubpartitionFileReader reader = new TestingHsSubpartitionFileReader();
+        CompletableFuture<Void> consumerStart = new CompletableFuture<>();
+        CompletableFuture<Void> readerFail = new CompletableFuture<>();
+        HsSubpartitionView subpartitionView =
+                new HsSubpartitionView(new NoOpBufferAvailablityListener());
+        reader.setFailConsumer(
+                (throwable) -> {
+                    try {
+                        readerFail.complete(null);
+                        consumerStart.get();
+                        // try to get view lock.
+                        subpartitionView.notifyDataAvailable();
+                    } catch (Exception e) {
+                        throw new RuntimeException(e);
+                    }
+                });
```

Comment:
This feels like whitebox testing.

Reply:
Fixed by override HsSubpartitionFileReaderImpl#fail method.